### DISCUSSION TITLE
SNlM0e value not found. Double-check __Secure-1PSID value or pass it as token='xxxxx'.

### DISCUSSION THREAD

**Discussion by hyuk-june** [dsdanielpark/Bard-API#138] [Q&A]
잘쓰고 있었는데 어느 순간부터 아래와 같은 에러가 발생합니다.
쿠키값을 다시 가져와 세팅해도 여전히 작동하지 않네요.
확인 좀 해주실 수 있으신가요?
감사합니다.

SNlM0e value not found. Double-check __Secure-1PSID value or pass it as token='xxxxx'.

**Comment by dsdanielpark** [comment]
안녕하세요. 지금 기능은 다시 체크해보았는데, 문제 없이 작동하는 것을 확인했습니다. 
API rate limiting이 걸리셨거나 일시적인 계정 제한일 수 있을 것 같습니다. 다른 계정의 쿠키 값으로 시도해보시거나 코랩에서 기능을 체크해보시길 권해드립니다. 다른 유저들에 따르면, 대부분 몇시간 뒤에 다시 정상작동하였다고 합니다. 감사합니다.

**Reply by hyuk-june** [reply]
아!  완전 로그아웃 하고 다시 쿠키 설정하니까 잘 되네요.
유용하게 잘 쓰고 있습니다.
좋은 자료 만들어 주셔서 정말 감사합니다.

**Comment by GvdDool** [comment]
Dear @dsdanielpark ,
First of all - congratulations on writing a well-working reverse-engineered API. 

I have no problem with the code, that's why I am tagging onto this discussion and not creating an issue, and the comment you made about "rate limiting is stuck, or it may be a temporary account limit." seems very relevant to my case.

To give some context:
I have written a small program to test the functionality when trying to scale up and get a feel for the accuracy of the program. The central question I am asking is:
'''Bard, can you give me all the countries in the sub-region: {sub_region}, as defined by the UN, in the format: "sub-region", "country name", and can you please make the list readable in Python code as a dictionary, with the name: sub_regions?''' 

This question works fine, but as there are 21 sub-regions (this was part of the test), the program fails randomly after 15 iterations, rarely making it to the end. I am on Windows, and I run this code on a Chrome browser in a CoLab environment while the cookie is obtained from a private window in the FireFox browser (to make refreshing the cookies easier). 

Again, the API is working well, and I get answers in the "correct" format, but on the tab in the FireFox browser, it often reads that the system has detected an unusual amount of traffic, and Google is asking to verify that I am not a bot. What kind of strategies are available to avoid being blocked in the loop, the cookie stays "fresh", and relaunching the loop sometimes runs again up to 15 requests, but when done too quickly after being blocked, the code will not work, so I am suspecting a  "temporary account limit".

**Comment by dsdanielpark** [comment]
Dear [GvdDool](https://github.com/GvdDool),

It is highly likely that there are restrictions imposed on the account and IP. While sending "sub-region" and "country name" as a dictionary can indeed be a relatively quick task, I am doubtful whether this alone can successfully bypass the restrictions. Unless using a proxy, merely specifying these in the request might not be sufficient to circumvent the limitations. If you happen to test and find that it works, please let me know, and I will add it accordingly.

**Reply by GvdDool** [reply]
Dear @dsdanielpark,
I am still investigating, and using a proxy to test the code is on my to-do list; what I noticed is that adding a random waiting time between requests is increasing the returns (probably this is circumventing one of their bot-detection tests), but this is not enough to let the loop (21 steps) complete without a cookie error.

**Reply by dsdanielpark** [reply]
(I have come to realize that I can write a message in the "reply" section in the response above.)

As you mentioned, incorporating a delay between requests using a random number for time.sleep might yield a slight effect. Applying a timeout of, for example, 60 seconds (or even longer) and using time.sleep with a value within 60 seconds could offer some benefits for longer tasks. While bypassing proxies might be an option, as you are aware, there are numerous simple methods to implement rate limiting. Therefore, detecting such behaviors from Google's end and imposing temporary account locks is not difficult.

I would greatly appreciate it if you could share your experimental results and if you could discover methods that others can use comfortably.

**Reply by GvdDool** [reply]
Sure, if I find a method that keeps the cookies fresh and doesn't put the (email) account in a temporary lock, I will share it with the community. Bard is a very nice experiment, and I would like to make efficient use of the (new) technology - which always involves more R&D than expected.

**Reply by GvdDool** [reply]
@dsdanielpark - quick follow-up question: I found this anecdotal phrase: "Meanwhile, Google Bard is limited to 1,000 tokens (or approximately 750 words) per session." and the word "session" made me think: can we, in code, force a new session, and if we can, how would we do this?  

Can we, for example, wrap this statement in a "with block": 
```
# Send an API request and get a response.
# test 2: time awareness and accuracy
input_text = '''"Hi Bard! I would like to know the current time in Paris, France, and the time difference with New York; what sources did you use?"'''
response = Bard(token1).get_answer(input_text)['content']
```
So that the session gets reitinated each time a request is made?

**Reply by dsdanielpark** [reply]
Since the session object already utilizes personalized targeting cookie values(snim0e value), simply re-declaring the session is an unnecessary task. However, if the logic to forcefully reset the cookie values is automated, there is a potential for circumventing the values by forcefully changing and resetting the cookies.

Nevertheless, even in this scenario, the frequent fluctuations in the value of the targeting cookie (snim0e) could lead Google to potentially impose temporary locks.

**Reply by dsdanielpark** [reply]
As mentioned earlier, you can forcibly delete the snim0e cookie value and then reassign it to token1. However, this approach is also likely to be easily detectable by Google.

Try to circumvent this by adjusting the timeout and the timing of resets in a highly randomized manner.

```python

input_text = '''"Hi Bard! I would like to know the current time in Paris, France, and the time difference with New York; what sources did you use?"'''

with Bard(token1) as bard:
    response = bard.get_answer(input_text)['content']

```

**Reply by GvdDool** [reply]
Thanks, I was already wondering what the "cookie values(snim0e value)" is doing, not an expert in API design and web development, but yes, that was my next idea: "forcefully reset the cookie values". This is not easy because the "browser_cookie3" is not picking up the __Secure-1PSID values (for me), as these are temporary values and don't make it to the SQLite db (as far as I know), so running "browser_cookie3" gives me only the permanent cookies, which makes picking up the __Secure-1PSID a little bit more challenging. 

I looked into running under selenium, but this is then blocked by the 2-way verification on the Google account to login to Bard; I was hoping to get the __Secure-1PSID when selenium was active by using: https://www.selenium.dev/documentation/webdriver/interactions/cookies/#get-all-cookies

**Reply by GvdDool** [reply]
@dsdanielpark , sorry for bugging you, but is the current class able to run in a "with block"? I am getting the following error:
<img width="424" alt="image" src="https://github.com/dsdanielpark/Bard-API/assets/54751004/5df5a116-750e-45d0-b1ce-82600ecc1f3d">

**Reply by GvdDool** [reply]
But I think I will have to try something else: 
I have implemented a wait time, average 3.8 seconds to stay within the 1000 requests an hour, and still get:
`Exception: Response status code is not 200. Response Status is 429`

> The HTTP 429 Too Many Requests response status code indicates the user has sent too many requests in a given amount of time ("rate limiting"). A Retry-After header might be included to this response indicating how long to wait before making a new request.

**Reply by GvdDool** [reply]
I was testing and hit the 20-minute "freshness" limit before I could get confirmation that there is a data limit (~7000) words, I managed two full runs (getting 7150 words in total) when the code broke with the error SNlM0e value not found. Double-check __Secure-1PSID value or pass it as token='xxxxx'.) Resetting the __Secure-1PSID, and rerunning gave me the HTTP 429 error in the second pass, with 75 successful requests; the total amount of words retrieved was 11050 in two sessions.

**Reply by GvdDool** [reply]
This said, the tool is very useful, but it is outside my skillset to make it scalable, which is a pity.
